### PR TITLE
Allow to consume from/to given timestamps

### PR DESCRIPTION
**Pull request by champtar**
Hi @edenhill 

now tested
With this PR we can now do `kafkacat -o s@<startts> -o e@<stopts> ...`

### CODE REVIEW

**Comment by Marcelovk** [comment]
:+1: 

I extended this here: https://github.com/champtar/kafkacat/pull/1

**Comment by champtar** [comment]
ping @edenhill

**Comment by champtar** [comment]
Hi @edenhill,

This PR is waiting for review for more than 8 months now, anything I can do to help ?

File: kafkacat.c [hunk 1]
```diff
@@ -628,6 +645,43 @@ static void kafkaconsumer_run (FILE *fp, char *const *topics, int topic_cnt) {
 }
 #endif
 
+/**
+ * Get offsets from conf.starts for consumer_run
+ */
+static int64_t* get_offsets(rd_kafka_metadata_topic_t *topic) {
```
Comment (edenhill) [inline kafkacat.c]: pointer goes on the symbol, not on the type, `int64_t *get_offsets..`.

File: kafkacat.c [hunk 2]
```diff
@@ -628,6 +645,43 @@ static void kafkaconsumer_run (FILE *fp, char *const *topics, int topic_cnt) {
...
+        int i;
+        rd_kafka_resp_err_t err;
+        rd_kafka_topic_partition_list_t *rktparlistp =
+                                           rd_kafka_topic_partition_list_new(1);
+
+        for (i = 0 ; i < (*topic).partition_cnt ; i++) {
+                int32_t partition = (*topic).partitions[i].id;
```
Comment (edenhill) [inline kafkacat.c]: `topic->partitions..`

File: kafkacat.c [hunk 3]
```diff
@@ -628,6 +645,43 @@ static void kafkaconsumer_run (FILE *fp, char *const *topics, int topic_cnt) {
...
+/**
+ * Get offsets from conf.starts for consumer_run
+ */
+static int64_t* get_offsets(rd_kafka_metadata_topic_t *topic) {
+        int i;
+        rd_kafka_resp_err_t err;
+        rd_kafka_topic_partition_list_t *rktparlistp =
+                                           rd_kafka_topic_partition_list_new(1);
+
+        for (i = 0 ; i < (*topic).partition_cnt ; i++) {
```
Comment (edenhill) [inline kafkacat.c]: `topic->`

File: kafkacat.c [hunk 4]
```diff
@@ -628,6 +645,43 @@ static void kafkaconsumer_run (FILE *fp, char *const *topics, int topic_cnt) {
 }
 #endif
 
+/**
+ * Get offsets from conf.starts for consumer_run
```
Comment (edenhill) [inline kafkacat.c]: startts

File: kafkacat.c [hunk 5]
```diff
@@ -628,6 +645,43 @@ static void kafkaconsumer_run (FILE *fp, char *const *topics, int topic_cnt) {
...
+                    partition)->offset = conf.startts;
+
+                if (conf.partition != RD_KAFKA_PARTITION_UA)
+                        break;
+        }
+        err = rd_kafka_offsets_for_times(conf.rk, rktparlistp, 10*1000);
+        if (err)
+                KC_FATAL("offsets_for_times failed: %s", rd_kafka_err2str(err));
+
+        int64_t *offsets = calloc(sizeof(int64_t), (*topic).partition_cnt);
```
Comment (edenhill) [inline kafkacat.c]: variabla definitions need to go at the top of the scope/block

File: kafkacat.c [hunk 6]
```diff
@@ -685,15 +739,22 @@ static void consumer_run (FILE *fp) {
...
 
                 if (conf.partition != RD_KAFKA_PARTITION_UA)
-                        part_eof_thres = 1;
+                        part_stop_thres = 1;
                 else
-                        part_eof_thres = metadata->topics[0].partition_cnt;
+                        part_stop_thres = metadata->topics[0].partition_cnt;
+        }
+
+        int64_t *offsets = NULL;
```
Comment (edenhill) [inline kafkacat.c]: move to head of scope/block

File: kafkacat.c (hunk 5 above)
Comment (edenhill) [reply kafkacat.c]: `topic->`

File: kafkacat.c (hunk 2 above)
Comment (champtar) [reply kafkacat.c]: I write C once a year, but I was creative this time

**Comment by champtar** [comment]
@edenhill fixed

**Comment by edenhill** [comment]
Thank you for your contribution!

**Comment by aoakeson** [comment]
@edenhill @champtar When using the command: 

`kafkacat -C -t mytopic -o s@1588331034000 e@1589022234000`

I am getting `Error: file/topic list only allowed in producer(-P)/kafkaconsumer(-G) mode`

```
Version:
kafkacat - Apache Kafka producer and consumer tool
https://github.com/edenhill/kafkacat
Copyright (c) 2014-2019, Magnus Edenhill
Version 1.5.0 (JSON, Avro, librdkafka 1.1.0 builtin.features=gzip,snappy,ssl,sasl,regex,lz4,sasl_gssapi,sasl_plain,sasl_scram,plugins,zstd,sasl_oauthbearer)
```

**Pull request merged**